File: easypermissions/src/main/java/pub/devrel/easypermissions/RationaleDialogFragmentCompat.java
package pub.devrel.easypermissions;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.annotation.StyleRes;
import androidx.fragment.app.FragmentManager;
import androidx.appcompat.app.AppCompatDialogFragment;

/**
 * {@link AppCompatDialogFragment} to display rationale for permission requests when the request
 * comes from a Fragment or Activity that can host a Fragment.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
public class RationaleDialogFragmentCompat extends AppCompatDialogFragment {

    public static final String TAG = "RationaleDialogFragmentCompat";

    private EasyPermissions.PermissionCallbacks mPermissionCallbacks;
    private EasyPermissions.RationaleCallbacks mRationaleCallbacks;

    public static RationaleDialogFragmentCompat newInstance(
            @NonNull String rationaleMsg,
            @NonNull String positiveButton,
            @NonNull String negativeButton,
            @StyleRes int theme,
            int requestCode,
            @NonNull String[] permissions) {

        // Create new Fragment
        RationaleDialogFragmentCompat dialogFragment = new RationaleDialogFragmentCompat();

        // Initialize configuration as arguments
        RationaleDialogConfig config = new RationaleDialogConfig(
                positiveButton, negativeButton, rationaleMsg, theme, requestCode, permissions);
        dialogFragment.setArguments(config.toBundle());

        return dialogFragment;
    }

    /**
     * Version of {@link #show(FragmentManager, String)} that no-ops when an IllegalStateException
     * would otherwise occur.
     */
    public void showAllowingStateLoss(FragmentManager manager, String tag) {
        if (manager.isStateSaved()) {
            return;
        }

        show(manager, tag);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (getParentFragment() != null) {
            if (getParentFragment() instanceof EasyPermissions.PermissionCallbacks) {
                mPermissionCallbacks = (EasyPermissions.PermissionCallbacks) getParentFragment();
            }
            if (getParentFragment() instanceof EasyPermissions.RationaleCallbacks){
                mRationaleCallbacks = (EasyPermissions.RationaleCallbacks) getParentFragment();
            }
        }

        if (context instanceof EasyPermissions.PermissionCallbacks) {
            mPermissionCallbacks = (EasyPermissions.PermissionCallbacks) context;
        }

        if (context instanceof EasyPermissions.RationaleCallbacks) {
            mRationaleCallbacks = (EasyPermissions.RationaleCallbacks) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mPermissionCallbacks = null;
        mRationaleCallbacks = null;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        // Rationale dialog should not be cancelable
        setCancelable(false);

        // Get config from arguments, create click listener
        RationaleDialogConfig config = new RationaleDialogConfig(getArguments());
        RationaleDialogClickListener clickListener =
                new RationaleDialogClickListener(this, config, mPermissionCallbacks, mRationaleCallbacks);

        // Create an AlertDialog
        return config.createSupportDialog(getContext(), clickListener);
    }
}

File: easypermissions/src/main/java/pub/devrel/easypermissions/dialogs/rationale/RationaleDialogFragmentCompat.kt
package pub.devrel.easypermissions.dialogs.rationale

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import androidx.annotation.RestrictTo
import androidx.annotation.StyleRes
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.fragment.app.FragmentManager
import pub.devrel.easypermissions.facade.EasyPermissions
import pub.devrel.easypermissions.models.PermissionRequest

/**
 * [AppCompatDialogFragment] to display rationale for permission requests when the request
 * comes from a Fragment or Activity that can host a Fragment.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
class RationaleDialogFragmentCompat : AppCompatDialogFragment() {

    private var mPermissionCallbacks: EasyPermissions.PermissionCallbacks? = null
    private var mRationaleCallbacks: EasyPermissions.RationaleCallbacks? = null

    /**
     * Version of [.show] that no-ops when an IllegalStateException
     * would otherwise occur.
     */
    fun showAllowingStateLoss(manager: FragmentManager, tag: String) {
        if (manager.isStateSaved) {
            return
        }

        show(manager, tag)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (parentFragment != null) {
            if (parentFragment is EasyPermissions.PermissionCallbacks) {
                mPermissionCallbacks = parentFragment as EasyPermissions.PermissionCallbacks?
            }
            if (parentFragment is EasyPermissions.RationaleCallbacks) {
                mRationaleCallbacks = parentFragment as EasyPermissions.RationaleCallbacks?
            }
        }

        if (context is EasyPermissions.PermissionCallbacks) {
            mPermissionCallbacks = context
        }

        if (context is EasyPermissions.RationaleCallbacks) {
            mRationaleCallbacks = context
        }
    }

    override fun onDetach() {
        super.onDetach()
        mPermissionCallbacks = null
        mRationaleCallbacks = null
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        // Rationale dialog should not be cancelable
        isCancelable = false

        // Get config from arguments, create click listener
        val config = RationaleDialogConfig(arguments!!)
        val clickListener =
            RationaleDialogClickListener(this, config, mPermissionCallbacks, mRationaleCallbacks)

        // Create an AlertDialog
        return config.createSupportDialog(context!!, clickListener)
    }

    companion object {

        val TAG = "RationaleDialogFragmentCompat"

        fun newInstance(permissionRequest: PermissionRequest): RationaleDialogFragmentCompat {

            // Create new Fragment
            val dialogFragment = RationaleDialogFragmentCompat()

            // Initialize configuration as arguments
            val config = RationaleDialogConfig(permissionRequest)
            dialogFragment.arguments = config.toBundle()

            return dialogFragment
        }
    }
}
